edisgo/flex_opt/reinforce_measures.py

```diff
@@ -725,3 +732,379 @@ def _replace_by_parallel_standard_lines(lines):
+        path = nx.shortest_path(G, station_node, node)
+
+        for first_node in relevant_lines.bus1.values:
+            if first_node in path:
+                paths[node] = path
+                first_nodes_feeders.setdefault(path[1], []).append(
+                    node  # first nodes and paths
+                )
+
+    lines_changes = {}
+    transformers_changes = {}
+    nodes_tb_relocated = {}  # nodes to be moved into the new grid
+
+    # note: The number of critical lines in the Lv grid can be more than 2. However,
+    # if the node_1_2 of the first feeder in the for loop is not the first node of the
+    # feeder, it will add data frames even though the following feeders only 1 node
+    # (node_1_2=first node of feeder). In this type of case,the number of critical lines
+    # should be evaluated for the feeders whose node_1_2 s are not the first node of the
+    # feeder. The first check should be done on the feeders that have fewer nodes.
+
+    first_nodes_feeders = dict(
+        sorted(
+            first_nodes_feeders.items(), key=lambda item: len(item[1]), reverse=False
+        )
+    )
+
+    count_inept = 0
+
+    for first_node, nodes_feeder in first_nodes_feeders.items():
+        # first line of the feeder
+        first_line = relevant_lines[relevant_lines.bus1 == first_node].index[0]
+
+        # the last node of the feeder
+        last_node = nodes_feeder[-1]
+
+        # the length of each line (the shortest path)
+        path_length_dict_tmp = dijkstra_shortest_path_length(
+            G, station_node, get_weight, target=last_node
+        )
+
+        # path does not include the nodes branching from the node on the main path
+        path = paths[last_node]
+
+        node_1_2 = next(
+            j
+            for j in path
+            if path_length_dict_tmp[j] >= path_length_dict_tmp[last_node] * 1 / 2
+        )
+
+        # if LVGrid: check if node_1_2 is outside a house
+        # and if not find next BranchTee outside the house
+        while (
+            ~np.isnan(grid.buses_df.loc[node_1_2].in_building)
+            and grid.buses_df.loc[node_1_2].in_building
+        ):
+            node_1_2 = path[path.index(node_1_2) - 1]
+            # break if node is station
+            if node_1_2 is path[0]:
+                raise NotImplementedError(
+                    f" {grid}==>{first_line} and following lines could not be "
+                    f"reinforced due to insufficient number of node in the feeder. "
+                    f"A method to handle such cases is not yet implemented."
+                )
+
+        # NOTE: If node_1_2 is a representative (meaning it is already directly
+        #  connected to the station) feeder cannot be split. Instead, every second
+        #  inept feeder is assigned to the new grid
+        if node_1_2 not in first_nodes_feeders or count_inept % 2 == 1:
+            nodes_tb_relocated[node_1_2] = nodes_feeder[nodes_feeder.index(node_1_2) :]
```

Comment:
This will only be correct, if nodes_feeder is sorted by distance to station, which I am not sure it is..